packages/core/core/src/AssetGraphBuilder.js

```diff
@@ -201,42 +199,49 @@ export default class AssetGraphBuilder extends EventEmitter {
     }
 
     this.rejected = new Map();
-    let lastQueueError;
-    for (let currPriorities of requestPriorities) {
-      if (!this.requestTracker.hasInvalidRequests()) {
-        break;
-      }
 
-      let promises = [];
-      for (let request of this.requestTracker.getInvalidRequests()) {
-        // $FlowFixMe
-        let assetGraphBuildRequest: AssetGraphBuildRequest = (request: any);
-        if (currPriorities.includes(request.type)) {
-          promises.push(this.queueRequest(assetGraphBuildRequest, {signal}));
-        }
-      }
-      if (lastQueueError) {
-        throw lastQueueError;
-      }
-      this.queue.run().catch(e => {
-        lastQueueError = e;
-      });
-      await Promise.all(promises);
+    let root = this.assetGraph.getRootNode();
+    if (!root) {
+      throw new Error('A root node is required to traverse');
     }
 
-    if (this.assetGraph.hasIncompleteNodes()) {
-      for (let id of this.assetGraph.incompleteNodeIds) {
-        this.processIncompleteAssetGraphNode(
-          nullthrows(this.assetGraph.getNode(id)),
+    let visited = new Set([root.id]);
+
+    const visit = node => {
```

Comment:
Is this basically `graph.bfs`?

Reply:
Yep :)

Reply:
Should we use that method? https://github.com/parcel-bundler/parcel/blob/5175a66b28cd2a6090b247204e550cc0264aabbe/packages/core/core/src/Graph.js#L458

Reply:
I guess I should have explained the differences more, one being that we wait for an async function to finish before visiting children. We also check to see if the children are deferred before visiting them and we revisit hasDeferred nodes if we encounter them again because that new connection may have changed the deferred status of the children. Wasn’t sure if it was worth adding those abstractions to the base graph. I guess if the visit function returned a promise we would know we should wait for it before visiting the children. I think we would also need a shouldVisit function that takes the visited set. I also ran into issues using the node in the set instead of the id because in certain cases the nodes were getting recreated, so I had to switch to using the id.

Reply:
That makes sense. I think it can stay here for now. If we come across another usecase it might make sense to combine.